### ISSUE TITLE
oxlint,oxfmt: support TypeScript ESM `.js` imports for actual `.ts` file from `.config.ts`

### ISSUE THREAD

**Issue by samchungy** [oxc-project/oxc#24285]
## Summary

Currently, importing other modules with a `.js` extension inside `oxfmt.config.ts` fails with `ERR_MODULE_NOT_FOUND`.

This makes it not so great to split a shared formatter config across multiple files which is fairly common pattern in monorepos or projects that share a base config.

Given a good majority of ESM repos won't be running with the `allowImportingTsExtensions` flag as it has other implications 😢 

https://github.com/samchungy/oxfmt-module-resolution

```ts
// oxfmt.config.ts
import { baseConfig } from './otherConfig.js'; // ❌ ERR_MODULE_NOT_FOUND

export default {
  ...baseConfig,
  semi: false,
};
```

## Root Cause

`oxfmt.config.ts` is loaded via a bare native Node.js `import()` call with no bundling or module resolution transformation (no `jiti`, `tsx`, `esbuild`, etc.). This means:

- Relative imports (e.g. `'./otherConfig.js'`) are resolved as literal file paths by Node's ESM loader - so a `.ts` source file referenced with a `.js` extension won't be found.
- Any missing package or unresolvable path will throw `ERR_MODULE_NOT_FOUND`, which is surfaced directly to the user with no actionable hint.

## Proposed Solution

Bundle or transform `oxfmt.config.ts` before executing it so that:

1. Relative imports within the config file are resolved and inlined correctly.
2. TypeScript-to-JS remapping works (`.js` extension → `.ts` source file).
3. Users can share a base config object across multiple files in a monorepo.

We could potentially use jiti for this?

## Workarounds (current)

- Keep `oxfmt.config.ts` entirely self-contained (no imports from other files).
- Use `.oxfmtrc.json` / `.oxfmtrc.jsonc` if no dynamic logic is needed.

**Comment by absidue** [comment]
The title and workarounds section are misleading, importing `.js` files works completely fine if you follow the ESM rules and only import files that actually exist on disk, so splitting your config is still possible.
What you are actually asking for is file extension guessing.

**Comment by samchungy** [comment]
Didn't mean to mislead - TypeScript by default kinda setup this pattern a while ago which makes it frustrating to deal with esepcially combined with ESM's requirement of requiring file extensions 🫠

edit: (I don't know what to call these sort of imports, happy for suggestions)

**Comment by leaysgur** [comment]
(re-titled a bit, since this also applies to oxlint, which uses the same shared config loader.)

I don't think we'll add a transform/bundling layer such as `jiti` here. It would add complexity, maintenance, and also runtime overhead + bundle size. Even Vite is also planning to move from bundled to native config loading: vitejs/vite#21546.

A custom loader hook could implement the remapping, but we'd then own non-standard resolution behavior and its maintenance. We'd prefer to follow native Node.js semantics.

Could you instead use an explicit `.ts` specifier with TypeScript 5.7’s `rewriteRelativeImportExtensions`? Unlike `allowImportingTsExtensions`, this works with emitting projects.

**Comment by samchungy** [comment]
> `rewriteRelativeImportExtensions`

TIL - I'll give it a shot thanks!

**Comment by samchungy** [comment]
We're going to pursue migrating our repos to use `.ts` extensions - thanks the suggestions
